resources/type-system/inference.md

```diff
@@ -970,6 +970,391 @@ with respect to `L` under constraints `C0`
+value of the variable `d`, and one which is the result of casting that value to
+the type `int`. The former artifact has static type `dynamic`; the latter has
+static type `int`. Since this document details precisely when and how coercions
+occur, and makes arguments about soundness of expressions that might involve
+coercions, it is useful to be precise by associating static types with the
+expression artifacts rather than the expressions themselves._
+
+## Soundness guarantees
+
+Along with the notion of _expression soundness_ described above, there are
+several other soundness guarantees made by Dart:
+
+- It is guaranteed that if a method, function, or constructor having return type
+  `T` is invoked, and the method completes with a value `v`, then `v` will be an
+  instance of `T` (with appropriate type substitutions, in the case of a generic
+  method). This guarantee is known as _return value soundness_.
+
+- It is guaranteed that if a method having static type `T` is torn off, then the
+  resulting value will be an instance of `T`. This guarantee is known as
+  _tear-off soundness_.
+
+- It is guaranteed that if a future is an instance of the type `Future<T>`, and
+  it completes with a value `v`, then `v` will be an instance of `T`. This
+  guarantee is known as _future soundness_.
+
+The rules below include informal sketches of a proof that each of the above
+soundness guarantees holds. These are non-normative, so they are typeset in
+_italics_.
+
+# Coercions
+
+_Coercion_ is a type inference step that transforms an expression artifact `m'`,
+and a desired static type `T`, into an expression artifact `m` whose static type
+is `T`.
+
+_Coercions are used in most situations where the existing spec calls for an
+assignability check._
+
+Coercion of an expression artifact `m'` to type `T` produces an expression
+artifact `m` with static type `T`, where `m` is determined as follows:
+
+- Let `T'` be the static type of `m'`.
+
+- Define `m` as follows:
+
+  - If `T' <: T`, then let `m` be an expression artifact whose runtime behavior
+    is as follows:
+
+    - Execute expression artifact `m'`, and let `o` be the resulting value. _By
+      execution soundness, `o` will be an instance of the type `T'`._
+
+    - `m` completes with the value `o`. _Expression soundenss follows from the
+      fact that since `T' <: T`, `o` must also be an instance of the type `T`._
+
+  - Otherwise, if `T'` is `dynamic`, then let `m` be an expression artifact
+    whose runtime behavior is as follows:
+
+    - Execute expression artifact `m'`, and let `o` be the resulting value. _By
+      execution soundness, `o` will be an instance of the type `T'`._
+
+    - If `o` is an instance of the type `T`, `m` completes with the value
+      `o`. _Expression soundness follows trivially._
+
+    - Otherwise, `m` completes with an exception. _Expression soundness follows
+      trivially._
```

Comment:
Likewise, maybe:

> _The value `o` isn't an instance of the coerced type, so the cast from `dynamic` fails and we throw an exception because there is no value to complete to that would preserve expression soundness._

Reply:
I reworked this one too:

> _In other words, the implicit cast from `dynamic` to `T_2` has failed. This trivially satisfies expression soundess, since expression soundness is only concerned with the value of the expression when it completes normally._